pagopa/pn-frontend
# Pn 2712 - FE - test di accessibilita e skipped

## Warning
The tests of `pn-validator` fail, for an issue I discussed with @AndreaCimini90 long ago, and which I thought it was settled. I wonder whether the code for that package is up-to-date.

## Short description
All the accessibility tests (i.e. the tests using `axe`) are moved to separated files. The scripts we use to run tests have changed in order to allow to run only "ordinary" tests, only accessibility tests, or all tests.

## List of changes proposed in this pull request
- For each test file including accessibility tests, a new file, which name ends with `.a11y.test.tsx` was created. The accessibility tests were moved to the new file.
- Whenever the "ordinary" and the accessibility tests shared initialization-related stuff, I created an additional file, which name ends with `test-utils.{ts|tsx}` to avoid duplication of such code.
- I also fixed some tests which were skipped, in the following files:
  - `pn-commons/src/components/SideMenu/__test__/SideMenuList.test.tsx`
  - `pn-personafisica-webapp/src/pages/__test__/AppStatus.page.test.tsx`
  - `pn-personafisica-webapp/src/pages/__test__/Deleghe.page.test.tsx`
- I re-enabled a test which was skipped for no apparent reason, in `pn-pf-webapp/src/pages/__test__/Contacts.page.test`. I guess this test was skipped by mistake, since the `.skip` was added in a commit whose message points at accessibility tests.
- In the cases in which there are still skipped tests, I added comments for further explaining the situation. These tests are in `pn-pa-webapp/src/pages/__test__/NotificationDetail.page.test.tsx` and `pn-pa-webapp/src/pages/components/NewNotification/__test__/Recipient.test.tsx`.
- In `pn-commons`, `pn-pa-webapp` and `pn-personafisica-webapp`, I modified the `test` script to avoid running the accessibility tests (so that only "ordinary" tests are included. I also added two new scripts, `test:a11y` tests only the accessibility tests (excluding the "ordinary" ones), and `test:all` runs all tests, whether "ordinary" or accessibility.
- The `ApiError` component now receives `apiId` as optional prop, and adding the value whenever passed to the `data-testid` of its outermost element.
- The code in `Deleghe.page.tsx` was slightly changed, erasing the strange ways in which some import were performed, since this particularity was introduced for a test which I changed and does not need it anymore.
- The subtitles in `Delegates`, `Delegators`  and `NotificationPayment` were "promoted" from `h6` to `h5`, in order to respect the [heading-order](https://dequeuniversity.com/rules/axe/4.6/heading-order) rule.
- A new Confluence page, namely https://pagopa.atlassian.net/wiki/spaces/PN/pages/594118729/Tests+automatici+di+accessibilit, describes how to write and run accessibility tests.

## How to test
- Run the `test`, `test:a11y`, `test:all` scripts in `pn-commons`, `pn-pa-webapp` and/or `pn-personafisica-webapp`.  
- Check the new Confluence page.
- Launch the PA and PF webapps, in the latter check in particular the "Deleghe" part.
